pkg/ebpf/netlink/socket.go

```diff
@@ -0,0 +1,255 @@
+	"golang.org/x/sys/unix"
+)
+
+var _ netlink.Socket = &Socket{}
+var errNotImplemented = errors.New("not implemented")
+
+// Socket is an implementation of netlink.Socket (github.com/mdlayher/netlink)
+// It's mostly a copy of the original implementation (netlink.conn) with a few optimizations:
+// * We don't MSG_PEEK as we use a pre-allocated buffer large enough to fit any netlink message;
+// * We use a buffer pool for the message data;
+// * We remove all the synchronization & go-channels cruft and bring it upstream in a cheaper/simpler way (Consumer)
+type Socket struct {
+	fd   *os.File
+	pid  uint32
+	conn syscall.RawConn
+
+	// A 32KB buffer which we use for polling the socket.
+	// Since a netlink message can't exceed that size
+	// (in *theory* they can be as large as 4GB (u32), but see link below)
+	// we can avoid message peeks and and essentially cut recvmsg syscalls by half
+	// which is currently a perf bottleneck in certain workloads.
+	// https://www.spinics.net/lists/netdev/msg431592.html
+	recvbuf []byte
+
+	// A pool of pre-allocated buffer objects which is used upstream
+	pool *bufferPool
+}
+
+func NewSocket(pool *bufferPool) (*Socket, error) {
+	fd, err := unix.Socket(
+		unix.AF_NETLINK,
+		unix.SOCK_RAW|unix.SOCK_CLOEXEC,
+		unix.NETLINK_NETFILTER,
+	)
+
+	if err != nil {
+		return nil, err
+	}
+
+	err = unix.SetNonblock(fd, true)
+	if err != nil {
+		syscall.Close(fd)
+		return nil, err
+	}
+
+	err = unix.Bind(fd, &unix.SockaddrNetlink{Family: unix.AF_NETLINK})
+	if err != nil {
+		syscall.Close(fd)
+		return nil, os.NewSyscallError("bind", err)
+	}
+
+	addr, err := unix.Getsockname(fd)
+	if err != nil {
+		syscall.Close(fd)
+		return nil, os.NewSyscallError("getsockname", err)
+	}
+
+	pid := addr.(*unix.SockaddrNetlink).Pid
+	file := os.NewFile(uintptr(fd), "netlink")
+
+	conn, err := file.SyscallConn()
+	if err != nil {
+		file.Close()
+		return nil, err
+	}
+
+	socket := &Socket{
+		fd:      file,
+		pid:     pid,
+		conn:    conn,
+		recvbuf: make([]byte, 32*1024),
+		pool:    pool,
+	}
+	return socket, nil
+}
+
+func (s *Socket) Send(m netlink.Message) error {
+	b, err := m.MarshalBinary()
+	if err != nil {
+		return err
+	}
+
+	addr := &unix.SockaddrNetlink{
+		Family: unix.AF_NETLINK,
+	}
+
+	ctrlErr := s.conn.Write(func(fd uintptr) bool {
+		err = unix.Sendmsg(int(fd), b, nil, addr, 0)
+		return ready(err)
+	})
+	if ctrlErr != nil {
+		return ctrlErr
+	}
+
+	return err
+}
+
+func (s *Socket) Receive() ([]netlink.Message, error) {
+	n, err := s.recvmsg(s.recvbuf, 0)
+	if err != nil {
+		return nil, os.NewSyscallError("recvmsg", err)
+	}
+
+	n = nlmsgAlign(n)
+
+	// Copy data to a buffer that can be used upstream
+	var b []byte
+	if n < msgBufferSize {
+		b = s.pool.Get()
+	} else {
+		b = make([]byte, n)
+	}
+	copy(b, s.recvbuf[:n])
+
+	raw, err := syscall.ParseNetlinkMessage(b[:n])
+	if err != nil {
+		return nil, err
+	}
+
+	msgs := make([]netlink.Message, 0, len(raw))
+	for _, r := range raw {
```

Comment:
Is the `b` buffer still referenced underneath these messages?  If not, it can be returned to the pool.

Reply:
It is: https://golang.org/src/syscall/netlink_linux.go

Reply:
Got it.  Thoughts on passing these back as is without allocating an intermediate slice?

Reply:
I'm very skeptical that this would help, to be honest. From the allocation profiles I took, `recvmsg` is responsible for ~1% of the allocations. If we don't group them in a slice I believe that would mean:
* More channel send/receive operations;
* It would be harder to do the buffer pool thing, since multiple messages scattered across multiple Event objects would reference the same chunk of memory;
* Error handling would get more complex (since one socket error would apply to multiple Events etc)

In case we do want to avoid these allocations I'd rather have a buffer pool for `[]netlink.Message`, or either change the existing pool to return something like 

```
type buffer struct {
   data   []byte
   msgs []netlink.Message
}
```
And the change would be mostly transparent outside the socket. But that might be overkill considering the 1% margin. What do you think?

Reply:
> More channel send/receive operations;

I had thought of just returning the existing `raw` slice, not sending them individually.  This should be the same number of send/receives, right ?

> or even change the existing pool to return something like

I think it would be good if `Socket.Receive()` returned that struct.  It would return the raw messages that it received along with the pooled buffer.

Reply:
Oh, I see. That makes more sense. I think it's a good idea but it would break the `netlink.Socket` interface, which I'd like to avoid at this point.